### PR TITLE
Golang: Make usage of contexts consistent

### LINKED ISSUE #1326: go/client should use context.Context consistently
Current status is that:
 - Some `LogClient` methods take a `context.Context`:
     - `GetSTHConsistency`
     - `GetProofByHash`
     - `LogClient.AddChainWithContext` (but there is no `AddPreChainWithContext`)
 - Some do not:
    - `GetSTH`
    - `AddPreChain`
    - `AddChain`
    - `AddJSON`

It would be nicer if everything took a `context.Context` parameter.

### PR DESCRIPTION
**Pull request by rolandshoemaker**
Remove the special method `client.AddChainWithContext` and make the following methods
take a `x/net/context.Context` as their first parameter:

* `GetSTH`
* `AddPreChain`
* `AddChain`
* `AddJSON`

Fixes #1326.

### CODE REVIEW

**Comment by daviddrysdale** [comment]
LGTM; Paul, are you OK with the API change?

File: go/client/ctclient/ctclient.go [hunk 1]
```diff
@@ -30,7 +31,7 @@ func signatureToString(signed *ct.DigitallySigned) string {
 }
 
 func getSTH(logClient *client.LogClient) {
-	sth, err := logClient.GetSTH()
+	sth, err := logClient.GetSTH(context.Background())
```
Comment (phad) [inline go/client/ctclient/ctclient.go]: nit: pass ctx to getSTH and have context.Background() used once only in the main().

File: go/client/ctclient/ctclient.go [hunk 2]
```diff
@@ -60,7 +61,7 @@ func addChain(logClient *client.LogClient) {
 		}
 	}
 
-	sct, err := logClient.AddChain(chain)
+	sct, err := logClient.AddChain(context.Background(), chain)
```
Comment (phad) [inline go/client/ctclient/ctclient.go]: same here.

File: go/client/logclient.go [hunk 3]
```diff
@@ -158,28 +158,22 @@ func (c *LogClient) addChainWithRetry(ctx context.Context, ctype ct.LogEntryType
 }
```
Comment (phad) [inline go/client/logclient.go]: (floating comment, github won't let me attach it in the right place!)
Now that we're passing ctx everywhere we should never be given a nil context.
So you should remove the 
```go
if ctx == nil {
  ctx = context.TODO()
}
```
lines in `addChainWithRetry` above.

File: go/preload/main/preload.go [hunk 4]
```diff
@@ -97,7 +98,7 @@ func certSubmitterJob(addedCerts chan<- *preload.AddedCert, log_client *client.L
 		chain := make([]ct.ASN1Cert, len(c.Chain)+1)
 		chain[0] = c.X509Cert.Raw
 		copy(chain[1:], c.Chain)
-		sct, err := log_client.AddChain(chain)
+		sct, err := log_client.AddChain(context.Background(), chain)
```
Comment (phad) [inline go/preload/main/preload.go]: same nit as before: pass `ctx` into `certSubmitterJob()` from `main()`

File: go/preload/main/preload.go [hunk 5]
```diff
@@ -115,7 +116,7 @@ func precertSubmitterJob(addedCerts chan<- *preload.AddedCert, log_client *clien
 	precerts <-chan *ct.LogEntry,
 	wg *sync.WaitGroup) {
 	for c := range precerts {
-		sct, err := log_client.AddPreChain(c.Chain)
+		sct, err := log_client.AddPreChain(context.Background(), c.Chain)
```
Comment (phad) [inline go/preload/main/preload.go]: as above

**Pull request merged**